test/testutils/validators.go

```diff
@@ -52,34 +52,68 @@ func ExpectLeaderSetExist(ctx context.Context, lws *leaderworkerset.LeaderWorker
 	gomega.Eventually(func() (bool, error) {
-		var headlessService corev1.Service
-		if err := k8sClient.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, &headlessService); err != nil {
+
+		// Always got the latest lws.
+		var lws leaderworkerset.LeaderWorkerSet
+		if err := k8sClient.Get(ctx, types.NamespacedName{Name: leaderWorkerSet.Name, Namespace: leaderWorkerSet.Namespace}, &lws); err != nil {
 			return false, err
 		}
-		// we expect exactly one service and it must be of the correct type.
-		if headlessService.ObjectMeta.Name != lws.Name {
-			return false, errors.New("service name mismatch")
-		}
-		if headlessService.ObjectMeta.Namespace != lws.Namespace {
-			return false, errors.New("service namespace mismatch")
+
+		var headlessService corev1.Service
+		var headlessServiceList corev1.ServiceList
+		if err := k8sClient.List(ctx, &headlessServiceList, client.InNamespace(lws.Namespace)); err != nil {
+			return false, err
 		}
-		if headlessService.Spec.ClusterIP != "None" {
-			return false, errors.New("service type mismatch")
+
+		if len(headlessServiceList.Items) != (numHeadlessServices) {
+			return false, fmt.Errorf("expected %d headless services, got %d", numHeadlessServices, len(headlessServiceList.Items))
 		}
-		if !headlessService.Spec.PublishNotReadyAddresses {
-			return false, errors.New("service publish not ready should be true")
+
+		if *lws.Spec.NetworkConfig.SubdomainPolicy == leaderworkerset.SubdomainShared {
+			if err := k8sClient.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, &headlessService); err != nil {
+				return false, err
+			}
+			return validateService(headlessService, &lws, lws.Name, map[string]string{leaderworkerset.SetNameLabelKey: lws.Name})
 		}
-		selector := headlessService.Spec.Selector
-		value, exists := selector["leaderworkerset.sigs.k8s.io/name"]
-		if !exists || value != lws.Name {
-			return false, errors.New("selector name incorrect")
+
+		for i := 0; i < int(*lws.Spec.Replicas); i++ {
+			if err := k8sClient.Get(ctx, types.NamespacedName{Name: fmt.Sprintf("%s-%s", lws.Name, strconv.Itoa(i)), Namespace: lws.Namespace}, &headlessService); err != nil {
+				return false, err
+			}
+			if _, err := validateService(headlessService, &lws, fmt.Sprintf("%s-%s", lws.Name, strconv.Itoa(i)), map[string]string{leaderworkerset.SetNameLabelKey: lws.Name, leaderworkerset.GroupIndexLabelKey: strconv.Itoa(i)}); err != nil {
+				return false, err
+			}
 		}
 		return true, nil
 	}, Timeout, Interval).Should(gomega.Equal(true))
 }
 
+func validateService(headlessService corev1.Service, lws *leaderworkerset.LeaderWorkerSet, serviceName string, wantSelector map[string]string) (bool, error) {
+	if headlessService.ObjectMeta.Name != serviceName {
+		return false, errors.New("service name mismatch")
+	}
+	if headlessService.ObjectMeta.Namespace != lws.Namespace {
+		return false, errors.New("service namespace mismatch")
+	}
```

Comment:
Checking the name and namespace is redundent, we already used them to get the object, can we please remove them.

Reply:
removed